core/domain/story_domain_test.py

```diff
@@ -915,6 +999,107 @@ def test_get_all_linked_exp_ids(self) -> None:
         self.assertEqual(
             self.story.story_contents.get_all_linked_exp_ids(),
             ['exp_1', 'exp_2', 'exp_3'])
+        with self.assertRaisesRegex(
+                ValueError, 'A node with exploration id exp_3 already exists.'):
+            self.story.update_node_exploration_id('node_4', 'exp_3')
+        self.story.update_node_exploration_id('node_3', 'exp_3')
+        self.assertEqual(
+            self.story.story_contents.get_all_linked_exp_ids(),
+            ['exp_1', 'exp_2', 'exp_3'])
+        self.story.update_node_exploration_id('node_3', 'exp_4')
+        self.assertEqual(
+            self.story.story_contents.get_all_linked_exp_ids(),
+            ['exp_1', 'exp_2', 'exp_4'])
+
+    def test_update_story_contents_from_model(self) -> None:
+        node_1: story_domain.StoryNodeDict = {
+            'id': 'node_1',
+            'thumbnail_filename': 'image.svg',
+            'thumbnail_bg_color': constants.ALLOWED_THUMBNAIL_BG_COLORS[
+                'chapter'][0],
+            'thumbnail_size_in_bytes': 21131,
+            'title': 'Title 1',
+            'description': 'Description 1',
+            'destination_node_ids': ['node_3'],
+            'acquired_skill_ids': [],
+            'prerequisite_skill_ids': [],
+            'outline': 'a',
+            'outline_is_finalized': False,
+            'exploration_id': 'exp_1'
+        }
+        story_contents_dict_1: story_domain.StoryContentsDict = {
+            'nodes': [node_1],
+            'initial_node_id': 'node_1',
+            'next_node_id': 'node_4'
+        }
+        version_dict: story_domain.VersionedStoryContentsDict = {
+            'schema_version': 0,
+            'story_contents': story_contents_dict_1
+        }
+
+        self.assertEqual(version_dict['schema_version'], 0)
+        story_domain.Story.update_story_contents_from_model(
+            version_dict, 1, 'node_1')
+        self.assertEqual(version_dict['schema_version'], 2)
+        self.assertEqual(
+            version_dict['story_contents']['nodes'][0]['thumbnail_filename'],
+            None)
```

Comment:
Instead of `assertEqual(..., None)`, it's more canonical to use `assertIsNone(...)`. Ditto elsewhere

Reply:
done